src/Raven.Server/Documents/RetireAttachmentsSender.cs

```diff
@@ -0,0 +1,417 @@
+
+                                using var hashSliceDisposable = _database.DocumentsStorage.AttachmentsStorage.RetiredAttachmentsStorage.ExtractHashSliceFromAttachmentId(context, doc.LowerId, out Slice hashSlice);
+                                var hash = hashSlice.ToString();
+
+                                var metadata = directUpload.GetObjectMetadata(string.Empty, hash);
+                                bool shouldUpload = true;
+                                long? attachmentLength = null;
+                                if (metadata != null)
+                                {
+                                    // The attachment already exists in the cloud, the file name is the hash so we can check if size matches to detect partial uploads
+                                    var objectSizeFromMetadata = directUpload.GetObjectSizeFromMetadata(metadata);
+                                    if (objectSizeFromMetadata.HasValue)
+                                    {
+                                        attachmentLength = AttachmentsStorage.GetAttachmentStreamLength(context, hashSlice);
+
+                                        // Only skip upload if sizes match exactly
+                                        shouldUpload = objectSizeFromMetadata != attachmentLength;
+                                    }
+                                }
+
+                                _token.ThrowIfCancellationRequested();
+
+                                if (shouldUpload == false)
+                                {
+                                    // the attachment already exists in the cloud, no need to upload it again
+                                    retired.Enqueue(doc);
+                                    continue;
+                                }
+
+                                // the attachment stream is disposed by the directUpload
+                                Stream attachmentStream;
+                                if (attachmentLength.HasValue == false)
+                                {
+                                    (attachmentStream, attachmentLength) = _database.DocumentsStorage.AttachmentsStorage.GetAttachmentStreamAndLength(context, hashSlice);
+                                }
+                                else
+                                {
+                                    attachmentStream = _database.DocumentsStorage.AttachmentsStorage.GetAttachmentStream(context, hashSlice);
+                                }
+
+                                if (attachmentStream == null)
+                                {
+                                    // attachment was deleted, need to remote it from retired tree
+                                    retired.Enqueue(doc);
+                                    continue;
+                                }
+
+                                if (await directUpload.WaitForFinishedTasksIfNeededAsync(duration, _token))
```

Comment:
this is how you are trying to avoid too many concurrent uploads - but I really get lost in the complexity.

We need to re-do all of this in proper, much smaller and focused API.
Using `async` propertly, etc.